ts/Extensions/Sonification/Instrument.ts

```diff
@@ -28,79 +28,655 @@ const {
+     *
+     * @function Highcharts.Instrument#stop
+     *
+     * @param {boolean} immediately
+     *        Whether to do the stop immediately or fade out.
+     *
+     * @param {Function} [onStopped]
+     *        Callback function to be called when the stop is completed.
+     *
+     * @param {*} [callbackData]
+     *        Data to send to the onEnd callback functions.
+     *
+     * @return {void}
+     */
+    public stop(
+        immediately: boolean,
+        onStopped?: Function,
+        callbackData?: unknown
+    ): void {
+        const instr = this,
+            reset = function (): void {
+                // Remove timeout reference
+                if (instr.stopOscillatorTimeout) {
+                    delete instr.stopOscillatorTimeout;
+                }
+                if (
+                    instr.oscillator &&
+                    instr.options.oscillator
+                ) {
+                    // The oscillator may have stopped in the meantime here, so
+                    // allow this function to fail if so.
+                    try {
+                        instr.oscillator.stop();
+                    } catch (e) {
+                        // silent error
+                    }
+                    if (instr.gainNode) {
+                        instr.oscillator.disconnect(instr.gainNode);
+                    }
+                    // We need a new oscillator in order to restart it
+                    instr.initOscillator(instr.options.oscillator);
+                }
+                // Done stopping, call the callback from the stop
+                if (onStopped) {
+                    onStopped(callbackData);
+                }
+                // Call the callback for the play we finished
+                if (instr.stopCallback) {
+                    instr.stopCallback(callbackData);
+                }
+            };
+
+        // Clear any existing timers
+        if (instr.playCallbackTimers.length) {
+            instr.clearPlayCallbackTimers();
         }
-        interface InstrumentPlayOptionsObject {
-            duration: number;
-            frequency: (number|Function);
-            maxFrequency?: number;
-            minFrequency?: number;
-            onEnd?: Function;
-            pan?: (number|Function);
-            volume?: (number|Function);
+        if (instr.stopTimeout) {
+            clearTimeout(instr.stopTimeout);
         }
-        interface OscillatorOptionsObject {
-            waveformShape?: InstrumentWaveform;
+        if (immediately) {
+            instr.setGain(0);
+            reset();
+        } else {
+            instr.mute();
+            // Stop the oscillator after the mute fade-out has finished
+            instr.stopOscillatorTimeout =
+                setTimeout(reset, H.sonification.fadeOutDuration + 100);
         }
     }
 }
 
+/* *
+ *
+ *  Class Namespace
+ *
+ * */
+
+namespace Instrument {
+
+    export let audioContext: AudioContext;
```

Comment:
This is not type safe. Change to optional static class property.

Reply:
Sorry, my mistake. The init function exits with error, if the context is missing. So this is fine.